### PR TITLE
Define `RegistryEntry` type and use it for dispatch

### PR DESCRIPTION
**Pull request by giordano**
Work-in-progress towards #20.

### CODE REVIEW

**Comment by giordano** [comment]
@ericphanson documentation still needs to be updated, but can you please do a first review of this pull request? :slightly_smiling_face:

**Comment by ericphanson** [comment]
So this does: `check(path::String)`, `check!(root, path::String)` (and vector variants), and `check(path::RegistryEntry)`, etc. I think that's pretty reasonable, and the implementation looks fine to me.

I think https://github.com/giordano/AnalyzeRegistry.jl/issues/20#issuecomment-803453046 could be an alternative, so that `check("Flux")` just works, for example.

**Comment by giordano** [comment]
Ok, done: now if the argument looks like a package name (`Base.isidentifier`) then it uses `find_package` to find the corresponding registry entry and check that one.

Now, what about the "verb"?  In https://github.com/giordano/AnalyzeRegistry.jl/issues/20#issuecomment-786264187 you suggested to use "check", but in https://github.com/giordano/AnalyzeRegistry.jl/issues/20#issuecomment-802838461 you seem to be back to "analyze"? :sweat_smile:

**Comment by ericphanson** [comment]
Cool! Looks good. Oh also, I recently learned about `Base.find_package` which I think is like ours but doesn’t restrict itself to General. It’s an internal function though so maybe we’re better off with ours.

For check vs analyze, true... CheckPackage sounds kind of weird to me, and it seems less clear what it does. But if the name is AnalyzePackage then it would make sense to call it analyze...

**Comment by giordano** [comment]
> Oh also, I recently learned about `Base.find_package` which I think is like ours but doesn’t restrict itself to General.

We don't do either, there is the `registry` keyword argument :smiley:  It looks to me that `Base.find_package` needs the package to be in the current environment though, which is much more limited

> For check vs analyze, true... CheckPackage sounds kind of weird to me, and it seems less clear what it does. But if the name is AnalyzePackage then it would make sense to call it analyze...

Ok, named back "analyze" :slightly_smiling_face:

**Comment by giordano** [comment]
Ok, [documentation](https://giordano.github.io/AnalyzeRegistry.jl/previews/PR31/) should now be up-to-date, and the PR is ready for final review :slightly_smiling_face:

File: docs/src/index.md [hunk 1]
```diff
@@ -1,11 +1,11 @@
 # AnalyzeRegistry.jl
 
-The main functionality of the package are the [`analyze`](@ref) and [`analyze_from_registry`](@ref) functions:
+The main functionality of the package is the [`analyze`](@ref) function:
 
 ```julia
 julia> using AnalyzeRegistry
 
-julia> analyze_from_registry(find_package("Flux"))
+julia> analyze(find_package("Flux"))
```
Comment (ericphanson) [inline docs/src/index.md:8]:
```suggestion
julia> analyze("Flux")
```
Maybe let's swap the order to show off the convenience?

File: docs/src/index.md [hunk 2]
```diff
@@ -24,19 +24,63 @@ Package Flux:
...
+is stored.  The function [`find_package`](@ref) gives you the
+[`RegistryEntry`](@ref) of a package in your local copy of any registry, by
+default the [General registry](https://github.com/JuliaRegistries/General).
 
 *NOTE*: the Git repository of the package will be cloned, in order to inspect
 its content.
 
+You can also pass the name of a package as argument to `analyze` instead of its
+path: in that case [`find_package`](@ref) will be automatically used to find its
+entry in the registry.
```
Comment (ericphanson) [inline docs/src/index.md:38]:
```suggestion
The argument is a string pointing towards a local path or the name of
a package in a locally-installed registry (the General registry is checked by default).

*NOTE*: the Git repository of the package will be cloned, in order to inspect
its content.

You can also pass a [`RegistryEntry`](@ref), a simple datastructure which points
to the directory of the package in the registry, where the file `Package.toml`
is stored.  The function [`find_package`](@ref) gives you the
[`RegistryEntry`](@ref) of a package in your local copy of any registry, by
default the [General registry](https://github.com/JuliaRegistries/General).
`find_package` is invoked automatically when you pass the name of a package.
```

File: docs/src/index.md [hunk 3]
```diff
@@ -24,19 +24,63 @@ Package Flux:
...
 
-You use the inplace version [`analyze_from_registry!`](@ref), e.g. as `analyze_from_registry!(root, find_package("Flux"))` to clone
+```julia
+julia> analyze("JuMP")
```
Comment (ericphanson) [inline docs/src/index.md:41]:
```suggestion
julia> analyze(find_package("JuMP"))
```

**Review by ericphanson — APPROVED** [review]
I like it! I suggested swapping the of the order of the examples in the docs, but I think either way is OK. Does the README need updating too?

**Comment by giordano** [comment]
> Does the README need updating too?

Yes, I completely forgot about it, thanks!

**Pull request merged**